File: src/quest_editor/actions/CreateEventAction.ts
import { Action } from "../../core/undo/Action";
import { QuestModel } from "../model/QuestModel";
import { QuestEventDagModel } from "../model/QuestEventDagModel";
import { QuestEventModel } from "../model/QuestEventModel";

export class CreateEventAction implements Action {
    readonly description: string;

    constructor(
        private readonly quest: QuestModel,
        private readonly event_dag: QuestEventDagModel,
        private readonly event: QuestEventModel,
        private readonly parent_event?: QuestEventModel,
    ) {
        this.description = `Add event ${event.id}`;
    }

    undo(): void {
        this.quest.remove_event(this.event_dag, this.event);
    }

    redo(): void {
        this.quest.add_event(this.event, this.parent_event ? [this.parent_event] : [], []);
    }
}

File: web/src/main/kotlin/world/phantasmal/web/questEditor/actions/CreateEventAction.kt
package world.phantasmal.web.questEditor.actions

import world.phantasmal.observable.change
import world.phantasmal.web.core.actions.Action
import world.phantasmal.web.questEditor.models.QuestEventModel
import world.phantasmal.web.questEditor.models.QuestModel

class CreateEventAction(
    private val setSelectedEvent: (QuestEventModel?) -> Unit,
    private val quest: QuestModel,
    private val index: Int,
    private val event: QuestEventModel,
) : Action {
    override val description: String = "Add event ${event.id.value}"

    override fun execute() {
        change {
            quest.addEvent(index, event)
            setSelectedEvent(event)
        }
    }

    override fun undo() {
        change {
            setSelectedEvent(null)
            quest.removeEvent(event)
        }
    }
}
